src/transpiling.rs

```diff
@@ -0,0 +1,810 @@
+    Program::Module(module) => module,
+    _ => unreachable!(),
+  };
+
+  Ok((source_file, module))
+}
+
+#[derive(Default, Clone)]
+struct DiagnosticCollector {
+  diagnostics_cell: Rc<RefCell<Vec<SwcDiagnostic>>>,
+}
+
+impl DiagnosticCollector {
+  pub fn into_handler(self) -> crate::swc::common::errors::Handler {
+    crate::swc::common::errors::Handler::with_emitter(
+      true,
+      false,
+      Box::new(self),
+    )
+  }
+}
+
+impl crate::swc::common::errors::Emitter for DiagnosticCollector {
+  fn emit(&mut self, db: &crate::swc::common::errors::DiagnosticBuilder<'_>) {
+    use std::ops::Deref;
+    self.diagnostics_cell.borrow_mut().push(db.deref().clone());
+  }
+}
+
+fn fold_program(
+  program: Program,
+  options: &EmitOptions,
+  source_map: Rc<SourceMap>,
+  comments: &SingleThreadedComments,
+  top_level_mark: Mark,
+) -> Result<Program> {
+  let jsx_pass = react::react(
+    source_map.clone(),
+    Some(comments),
+    react::Options {
+      pragma: options.jsx_factory.clone(),
+      pragma_frag: options.jsx_fragment_factory.clone(),
+      // this will use `Object.assign()` instead of the `_extends` helper
+      // when spreading props.
+      use_builtins: true,
+      runtime: if options.jsx_automatic {
+        Some(react::Runtime::Automatic)
+      } else {
+        None
+      },
+      development: options.jsx_development,
+      import_source: options.jsx_import_source.clone().unwrap_or_default(),
+      ..Default::default()
+    },
+    top_level_mark,
+  );
+  let mut passes = chain!(
+    Optional::new(
+      transforms::ImportDeclsToVarDeclsFolder,
+      options.repl_imports
+    ),
+    Optional::new(transforms::StripExportsFolder, options.repl_imports),
+    proposals::decorators::decorators(proposals::decorators::Config {
+      legacy: true,
+      emit_metadata: options.emit_metadata
+    }),
+    helpers::inject_helpers(),
+    resolver_with_mark(top_level_mark),
+    Optional::new(
+      typescript::strip::strip_with_config(
+        strip_config_from_emit_options(options),
+        top_level_mark
+      ),
+      !options.transform_jsx
+    ),
+    Optional::new(
+      typescript::strip::strip_with_jsx(
+        source_map.clone(),
+        strip_config_from_emit_options(options),
+        comments,
+        top_level_mark
+      ),
+      options.transform_jsx
+    ),
+    Optional::new(jsx_pass, options.transform_jsx),
+    fixer(Some(comments)),
+    hygiene(),
+  );
+
+  let emitter = DiagnosticCollector::default();
+  let diagnostics_cell = emitter.diagnostics_cell.clone();
+  let handler = emitter.into_handler();
+  let result = crate::swc::utils::HANDLER.set(&handler, || {
+    helpers::HELPERS.set(&helpers::Helpers::new(false), || {
+      program.fold_with(&mut passes)
+    })
+  });
+
+  let diagnostics = diagnostics_cell.borrow();
+  ensure_no_fatal_swc_diagnostics(&source_map, diagnostics.iter())?;
+  Ok(result)
+}
+
+fn strip_config_from_emit_options(
```

Comment:
discussion: this seemed oddly named to me and didn't have any comments, finally figured out what it was but only by reading the code... maybe `to_typescript_strip_config()` plus a comment instead?

Reply:
Yeah, it's a weird name and was thinking that too, but didn't touch the code much when moving it here. I'll change this to be a method off `EmitOptions`